### PR TITLE
Add support for service account authentication

### PR DESCRIPTION
**Pull request by BarakSason**
Adding support for service account authentication using the following
lib https://github.com/libopenstorage/secrets

Signed-off-by: Barak Sason Rofman <sason922@gmail.com>

### CODE REVIEW

File: pkg/system/phase2_creating.go [hunk 1]
```diff
@@ -723,15 +725,21 @@ func (r *Reconciler) ReconcileRootSecret() error {
...
 		if err := util.ValidateConnectionDetails(connectionDetails, authTokenSecretName, options.Namespace); err != nil {
 			return fmt.Errorf("ReconcileRootSecret: could not get/put key in external KMS: external kms connection details validation failed: %q", err)
 		}
 		kmsProvider := connectionDetails["KMS_PROVIDER"]
 		if util.IsVaultKMS(kmsProvider) {
-			keySecretName := "rootkeyb64-" + string(r.NooBaa.ObjectMeta.UID)
-			// reconcile root master key externally (vault)
-			c, err := util.InitVaultClient(connectionDetails, authTokenSecretName, options.Namespace)
-			if err == nil {
+			authMethod := connectionDetails["AUTH_METHOD"]
```
Comment (romayalon) [inline pkg/system/phase2_creating.go:734]: is this AUTH_METHOD really exist? someone should inject it to noobaa CR

File: pkg/system/phase2_creating.go [hunk 2]
```diff
@@ -757,8 +765,19 @@ func (r *Reconciler) ReconcileRootSecret() error {
 					}
 					return fmt.Errorf("ReconcileRootSecret: Error put secret in vault: %+v", err)
 				}
+			} else if(authMethod == "serviceaccount") {
+				config := make(map[string]interface{})
```
Comment (romayalon) [inline pkg/system/phase2_creating.go:769]: can you explain how this authentication works? config is empty map

File: pkg/system/phase2_creating.go [hunk 3]
```diff
@@ -723,15 +725,21 @@ func (r *Reconciler) ReconcileRootSecret() error {
...
 		}
 		kmsProvider := connectionDetails["KMS_PROVIDER"]
 		if util.IsVaultKMS(kmsProvider) {
-			keySecretName := "rootkeyb64-" + string(r.NooBaa.ObjectMeta.UID)
-			// reconcile root master key externally (vault)
-			c, err := util.InitVaultClient(connectionDetails, authTokenSecretName, options.Namespace)
-			if err == nil {
+			authMethod := connectionDetails["AUTH_METHOD"]
+			if(authMethod == "token") {
+				keySecretName := "rootkeyb64-" + string(r.NooBaa.ObjectMeta.UID)
```
Comment (romayalon) [inline pkg/system/phase2_creating.go:736]: in the old token auth method the key is stored under NOOBAA_ROOT_SECRET_PATH, in here you removed it. I think we should keep it in the same place in this auth mehod too.

File: pkg/system/phase2_creating.go [hunk 4]
```diff
@@ -769,6 +788,38 @@ func (r *Reconciler) ReconcileRootSecret() error {
 	return nil
 }
 
+func (r *Reconciler) putSecret(v secrets.Secrets, secretName, secretValue string, keyContext map[string]string) error {
```
Comment (romayalon) [inline pkg/system/phase2_creating.go:791]: please move it to util.go

File: pkg/system/phase2_creating.go [hunk 5]
```diff
@@ -769,6 +788,38 @@ func (r *Reconciler) ReconcileRootSecret() error {
...
+
+func getSecret(v secrets.Secrets, secretName string, keyContext map[string]string) (string, error) {
+	s, err := v.GetSecret(secretName, keyContext)
+	if err != nil {
+		return "", err
+	}
+
+	return s[secretName].(string), nil
+}
+
```
Comment (romayalon) [inline pkg/system/phase2_creating.go:822]: delete secret is missing currently right? couldn't find it

File: pkg/system/phase2_creating.go [hunk 6]
```diff
@@ -769,6 +788,38 @@ func (r *Reconciler) ReconcileRootSecret() error {
...
+	if key != "" {
+		r.SecretRootMasterKey.StringData["cipher_key_b64"] = key
+		return fmt.Errorf("key %q already exists in vault", secretName)
+	}
+
+	// Build Secret
+	data := make(map[string]interface{})
+	data[secretName] = secretValue
+
+	err = v.PutSecret(secretName, data, keyContext)
```
Comment (romayalon) [inline pkg/system/phase2_creating.go:806]: will it work for kv v2? please test it if you didn't

File: pkg/system/phase2_creating.go [hunk 7]
```diff
@@ -757,8 +765,19 @@ func (r *Reconciler) ReconcileRootSecret() error {
...
 				}
+			} else if(authMethod == "serviceaccount") {
+				config := make(map[string]interface{})
+				v, err := vault.New(config)
+				if err != nil {
+					return fmt.Errorf("failed to initialize vault secret store %+v", err)
+				}
+
+				keySecretName := "rootkeyb64-" + string(r.NooBaa.ObjectMeta.UID)
+				err = r.putSecret(v, keySecretName, r.SecretRootMasterKey.StringData["cipher_key_b64"], nil)
```
Comment (romayalon) [inline pkg/system/phase2_creating.go:776]: have you tested it on non-dev vault server?

**Comment by nimrod-becker** [comment]
Superseded by Alex's PR for the KSM Service Auth and HPCS support
 #797

**Pull request closed without merging**